[admin] CMI 관리자 페이지 UI 만들기

soonitoon opened the pull request:
# CMI 관리자 페이지 UI 만들기

## 👀 이슈

resolve #38

## 📌 개요

충림이 스크래퍼의 각 시나리오를 관리할 수 있는 관리자 페이지의 UI를 만들었습니다. 각 시나리오들은 그룹별로 메인 화면에 보이며, 시나리오 상태에 따라 필터링하여 볼 수도 있습니다.

## 👩‍💻 작업 사항

- 관리자 페이지를 왼쪽 사이드바와 메인 화면으로 구성했습니다.
- 메인 화면에는 주제목과 각 그룹별 제목을 표시했습니다.
- 각 시나리오 카드에는 시나리오 제목, 부제목, 태그를 표시하고 시나리오의 상태를 알 수 있는 신호등을 추가했습니다.
- 향후 시나리오별 세부 페이지 이동을 위해 각 시나리오에 마우스 커서를 올리면 파란 경계선이 시나리오 카드에 생기게 했습니다.

## ✅ 참고 사항

### 스크린샷

<img width="1552" alt="스크린샷 2021-11-17 오후 5 24 15" src="https://user-images.githubusercontent.com/71015915/142162806-469412a8-2a74-4ff8-a180-0d2c0ac3810e.png">

### `border` 표시 GIF

![border](https://user-images.githubusercontent.com/71015915/142162585-48259b3e-35d7-4e3d-8ca4-b3f925bc1a45.gif)

### 시나리오 상태에 따른 필터링 GIF

![statusFiltering](https://user-images.githubusercontent.com/71015915/142162602-2b334492-678b-4445-82db-e4f3d93b4198.gif)

soonitoon committed:
feat: Add style to admin page (#38)

nohgijin commented on packages/admin/src/__mocks__/noticeScenarios.ts:
```diff
@@ -1,46 +1,84 @@
   {
     id: 1,
-    title: "소프트웨어학과",
+    titleText: "소프트웨어학과",
```
왜 titleText로 바꾸셨나요? title이 좋을것 같아서...

jaryapp replied:
그러게여 저두 궁금 👀

soonitoon replied:
시나리오 카드 타이틀에 적용할 CSS 해쉬 변수명이 같은 title이어서 구분을 위해 바꿨는데, 해쉬 변수명을 바꾸거나 해쉬를 사용하지 않고 h2에 직접 스타일을 적용해야겠습니다.

jaryapp replied:
아 그래서 그러셨던거군여 그 스타일에서 값 가져올때는 비구조화 할당을 쓰지않구 `className`에 `style.title` 이런식으로 주시면 좋을것 같아여

soonitoon replied:
아아 그 방법을 생각 못 했네요 😂

nohgijin commented on packages/admin/src/components/Scenario/Card/index.tsx:
```diff
@@ -1,32 +1,55 @@
-    status: string;
-    tags: Array<string>;
-  };
+interface ScenarioCardConfig {
```
오늘 정한것 처럼 Props로 써주시는게 좋을 것 같아요!

soonitoon replied:
앞으로는 Props로 통일해서 사용하겠습니다!

nohgijin commented on packages/admin/src/components/Scenario/Card/index.tsx:
```diff
@@ -1,32 +1,55 @@
+  // TODO: ESLint 규칙에 추가하기
+  // eslint-disable-next-line default-case
+  switch (status) {
+    case sidebarMenus[1]:
```
status가 sidebarMenus에 들어있는 것만 올 수 있나요?
이 경우때문에 eslint규칙을 추가하기에는 다른 거에서도 이 기능을 사용을 못하는 거여서
default를 넣어주면 좋을 것 같아요!

jaryapp replied:
`sidebarMenus` 는 `enum`으로 관리하는게 좋을것 같아요

soonitoon replied:
status에 들어갈 수 있는 값은  jess와 시나리오의 데이터 구조에 대해 회의하면서 구체적으로 정해야 할 것 같습니다. 만약 status가 가질 수 있는 값이 지금처럼 세 가지 경우로 유지된다면 세 case 중 하나를 default로 바꾸겠습니다.

nohgijin commented on packages/admin/src/pages/Home/style.ts:
```diff
@@ -0,0 +1,20 @@
+      font-weight: 400;
+      font-size: 3rem;
+      margin-left: 5%;
+      margin-top: 15vh;
+      margin-bottom: 10vh;
```
vh를 쓰신 특별한 이유가 있나용?
다른건 다 rem으로 하셔서 ㅎㅎ

soonitoon replied:
앗 PR 날리기 전에 단위 통일한다면서 깜빡하고 있었네요ㅠ 저 당시에는 top margin이  브라우저 창 높이에 따라 일정 비율을 유지하게 만들고 싶어서 사용했습니다.

nohgijin commented on packages/admin/src/components/Scenario/Card/index.tsx:
```diff
@@ -1,32 +1,55 @@
 }
 
 export default function ScenarioCard({
   className,
```
아 이건 수니툰이 한게 아닌것 같은데
여기 className 빼면 오류가 날까요?
만약 오류가 안난다면 지워주셔도 될 듯 합니당...!

soonitoon replied:
이 부분은 jess에게 설명을 들은 것 같아요. 시나리오 카드는 다른 페이지에서도 재사용할 수 있도록 기본 스타일만 작성해두고, 해당 카드를 쓰는 상위 컴포넌트에서 입맛에 맞게 시나리오 카드에 스타일을 덧입히는 구조인 것 같습니다. 시나리오 카드 리스트 컴포넌트에서 시나리오 카드 배치에 대한 스타일을 각각의 시나리오 카드에 적용하기 위해서 CSS 해쉬 클래스 네임을 시나리오 카드에게 프롭으로 넘기는 구조가 아닐까 싶습니다...!

jaryapp replied:
넹 맞습니다 여기서 시나리오 카드그룹에서 카드 레이아웃 잡는 용도로 `className`은 필요해여~

nohgijin commented on packages/admin/src/utils/menuContext.ts:
```diff
@@ -0,0 +1,18 @@
+  setContext: (status: string) => {},
+};
+
+export const menuContext = createContext(defaultContext);
```
이런 컨텍스트는 보통 대문자로 쓰곤 합니다
https://ko.reactjs.org/docs/context.html#gatsby-focus-wrapper
타입이나 인터페이스두요~

soonitoon replied:
PR 전에 코드를 확인한다고 확인했는데 실수가 군데군데 보이네요ㅠ 수정하겠습니다!

nohgijin reviewed (commented):
수고하셨습니다!
혹시 캡쳐하실때 창 전체가 캡쳐되는 거는 어떤 툴을 쓰시나용 ?

soonitoon commented:
> 수고하셨습니다! 혹시 캡쳐하실때 창 전체가 캡쳐되는 거는 어떤 툴을 쓰시나용 ?

맥 OS에서 `shift + cmd + 5` 누르시면 창 전체 캡쳐할 수 있는 옵션이 나옵니다!

jaryapp commented on packages/admin/src/__mocks__/index.ts:
```diff
@@ -1 +1,4 @@
-export { default as noticeScenariosMocks } from "./noticeScenarios";
+import noticeGroupsMocks from "./noticeGroups";
+import noticeScenariosMocks from "./noticeScenarios";
+
+export { noticeScenariosMocks, noticeGroupsMocks };
```
mock 데이터 만들어서 작업하신거 최고입니다요~ 👍

soonitoon replied:
감사합니다! 🙏

jaryapp commented on packages/admin/src/components/Navigation/index.jsx:
```diff
@@ -0,0 +1,29 @@
+            <button
+              type="button"
+              onClick={() => setContext(item)}
+              className={cx(insideBtn, item === status ? activated : "")}
```
```suggestion
              className={cx(insideBtn, {activated: item === status} )}
```
이렇게 하시면 좋을것 같습니다

soonitoon replied:
이모션에서 제공하는 `cx()`의 사용법에 대해 제대로 이해하지 못하고 작성했던 것 같습니다. 알려주신 코드로 수정하겠습니다~

soonitoon replied:
혹시 이렇게 쓰는 게 맞을까요? 에러가 나서 공식문서를 찾아보니 대괄호가 있어서요!

```javascript
className={cx(style.insideBtn, { [style.activated]: item === status })
```

jaryapp replied:
대괄호가 맞는것 같아여 ㅎㅎㅎ...

jaryapp reviewed (commented):
수니툰 UI/UX 정말 맘에드네요ㅎㅎ 고생많으셨어요~!

jaryapp commented:
retest this please

jaryapp commented:
retest this please

jaryapp commented:
retest this please

jaryapp commented:
retest this please

jaryapp commented:
retest this please

kingyong9169 commented on packages/admin/src/utils/menuContext.ts:
```diff
@@ -0,0 +1,18 @@
+
+  // TODO: 아래의 ESLint 규칙 config 파일에 적용하기
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  setContext: (status: string) => {},
```
저도 수니툰 코드보고 처음 알게 되어서 공유합니다!
[this, function, 화살표 함수](https://wormwlrm.github.io/2019/03/04/You-should-know-JavaScript-this.html)
[객체에서 함수 선언시 function, 화살표 함수 차이](https://velog.io/@zwonlala/%EA%B0%9D%EC%B2%B4-%EC%95%88%EC%97%90-%ED%95%A8%EC%88%98-%EB%84%A3%EA%B8%B0)

정확히는 모르겠지만 useContext를 사용해서 setContext 호출할 때 별 문제가 없었던것 같습니다! 다시 공부해보고 말씀드리겠습니다!

soonitoon replied:
코드 리뷰 다시 달아주셔서 감사합니다ㅠ 앞으로는 브랜치 푸쉬할 때 컨벤션 따랐는지 다시 한 번 확인해야겠습니다

kingyong9169 commented on packages/admin/src/pages/Home/index.tsx:
```diff
@@ -0,0 +1,29 @@
+import getStyle from "./style";
+
+export default function Home() {
+  const [status, setStatus] = useState(sidebarMenus[0]);
```
린트 적용이 저처럼 나타나는것 같네요ㅠ 제스랑 저랑 같이 확인하시면 좋을 것 같습니다.

soonitoon replied:
혹시 저희가 따르는 포메팅과 다른가요? 저는 저장할 때마다 포메팅이 적용되길래 잘못된 줄도 모르고 있었습니다ㅠ

kingyong9169 replied:
아까 트레저가 말씀해주신 `[ status, setStatus ]` 이 부분입니다!

kingyong9169 reviewed (commented):
수고하셨습니다 수니툰~

jaryapp commented:
retest this please

soonitoon committed:
refactor: Clean code by code reviews (#50)

- `titleText` `title`로 다시 바꾸기.
  - 시나리오 객체에서 `titleText`를 `title`로 되돌림.
  - `getStyle()`에서 비구조화를 통해 가져오는 모든 코드를 비구조화를 사용하지 않는 방식으로 수정함(변수명이 겹쳐 불필요한 변수명을 사용하게 되는 상황을 방지하고자 함.).
- 한 파일 안에서만 사용하는 프롭 타입을 정의하는 경우 타입 이름을 `Props`로 통일.
- 타입 정의 시 `interface` 를 모두 `type`으로 바꿈.
- 타입 이름과 React Context 이름은 파스칼 표기법을 따라 첫 글자를 대문자로 바꿈.
- 사이드바에 표시되는 메뉴들을 타입스크립트의 `enum`으로 정의함. 이후 네비게이션 컴포넌트에서 enum에 정의된 각 메뉴들을 리스트로 직접 만들어서 사용하도록 함.
- 브라우저의 폰트 크기 설정을 존중할 수 있는 `rem` 단위 사용하기([관련 내용 링크](https://brunch.co.kr/@clay1987/170)).
- 이모션의 `cx` 함수 사용할 때 삼항연산자 대신 공식문서에서 권장하는 방법으로 조건부 클래스 네임 사용하기.
- NAVER CSS 속성 순서 컨벤션을 참고하여 CSS 속성 정리.
